File: backend/app/api/v1/endpoints/customers.py
from datetime import UTC, date, datetime
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy import ColumnElement, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.dependencies import get_db
from app.common.exceptions import NotFoundError, ValidationAppError
from app.common.phone import InvalidPhoneNumberError, normalize_phone
from app.database.models.customer import Customer, CustomerStatus, CustomerType
from app.database.models.customer_profile_token import CustomerProfileToken
from app.modules.customers.schemas import (
    CustomerCreate,
    CustomerCreateResponse,
    CustomerProfileTokenIssued,
    CustomerProfileTokenResponse,
    CustomerRead,
    CustomersListResponse,
    CustomersMeta,
    CustomerStats,
    CustomerStatsResponse,
    CustomerUpdate,
    UpcomingBirthday,
    UpcomingBirthdaysResponse,
)

router = APIRouter()

_SORTABLE_COLUMNS: dict[str, ColumnElement] = {
    "name": Customer.name,
    "total_spent": Customer.total_spent,
    "created_at": Customer.created_at,
}


def _days_until_next_birthday(dob: date, today: date) -> int:
    try:
        next_birthday = dob.replace(year=today.year)
    except ValueError:
        # dob is Feb 29 and this year isn't a leap year.
        next_birthday = date(today.year, 2, 28)

    if next_birthday < today:
        try:
            next_birthday = next_birthday.replace(year=today.year + 1)
        except ValueError:
            next_birthday = date(today.year + 1, 2, 28)

    return (next_birthday - today).days


async def _get_customer_or_404(db: AsyncSession, customer_id: UUID) -> Customer:
    customer = (
        await db.execute(select(Customer).where(Customer.public_id == customer_id))
    ).scalar_one_or_none()
    if customer is None:
        raise NotFoundError("Customer not found")
    return customer


@router.post(
    "", response_model=CustomerCreateResponse, status_code=status.HTTP_201_CREATED
)
async def create_customer(
    payload: CustomerCreate, db: AsyncSession = Depends(get_db)
) -> CustomerCreateResponse:
    try:
        normalized_phone = normalize_phone(payload.phone)
    except InvalidPhoneNumberError as exc:
        raise ValidationAppError(str(exc)) from exc

    existing = await db.execute(
        select(Customer).where(Customer.normalized_phone == normalized_phone)
    )
    if existing.scalar_one_or_none() is not None:
        raise ValidationAppError("A customer with this phone number already exists")

    customer = Customer(
        name=payload.name,
        phone=payload.phone.strip(),
        normalized_phone=normalized_phone,
        email=payload.email,
        address=payload.address,
        date_of_birth=payload.date_of_birth,
        is_vip=payload.is_vip,
    )
    db.add(customer)
    await db.commit()
    await db.refresh(customer)

    return CustomerCreateResponse(data=CustomerRead.model_validate(customer))


@router.get("", response_model=CustomersListResponse)
async def list_customers(
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = Query(None, description="Matches name, phone, or email"),
    status: str | None = Query(None),
    is_vip: bool | None = Query(None),
    customer_type: str | None = Query(None, description="GENERAL, VIP, or VVIP"),
    sort_by: str | None = Query(None),
    sort_dir: Literal["asc", "desc"] = Query("asc"),
) -> CustomersListResponse:
    filters: list[ColumnElement] = []

    search = (search or "").strip()
    if search:
        pattern = f"%{search}%"
        filters.append(
            or_(
                Customer.name.ilike(pattern),
                Customer.phone.ilike(pattern),
                Customer.email.ilike(pattern),
            )
        )

    if status and status != "all":
        filters.append(Customer.status == status)

    if is_vip is not None:
        filters.append(Customer.is_vip == is_vip)

    if customer_type and customer_type != "all":
        valid_types = {member.value for member in CustomerType}
        if customer_type not in valid_types:
            raise ValidationAppError(
                f"Invalid customer_type {customer_type!r}; expected one of {sorted(valid_types)}"
            )
        filters.append(Customer.customer_type == customer_type)

    count_query = select(func.count()).select_from(Customer)
    list_query = select(Customer)
    for condition in filters:
        count_query = count_query.where(condition)
        list_query = list_query.where(condition)

    total = (await db.execute(count_query)).scalar_one()

    order_column = _SORTABLE_COLUMNS.get(sort_by or "", Customer.name)
    order_clause = order_column.desc() if sort_dir == "desc" else order_column.asc()
    list_query = list_query.order_by(order_clause).offset((page - 1) * page_size).limit(page_size)

    customers = (await db.execute(list_query)).scalars().all()
    total_pages = max(1, -(-total // page_size))

    return CustomersListResponse(
        data=[CustomerRead.model_validate(customer) for customer in customers],
        meta=CustomersMeta(page=page, page_size=page_size, total=total, total_pages=total_pages),
    )


@router.get("/stats", response_model=CustomerStatsResponse)
async def get_customer_stats(db: AsyncSession = Depends(get_db)) -> CustomerStatsResponse:
    """Descriptive counts for the dashboard overview — reads existing columns
    only (is_vip, total_spent, date_of_birth); no VIP/birthday business logic
    lives here."""
    total_customers = (await db.execute(select(func.count()).select_from(Customer))).scalar_one()

    vip_customers = (
        await db.execute(
            select(func.count()).select_from(Customer).where(Customer.is_vip.is_(True))
        )
    ).scalar_one()

    total_revenue = (
        await db.execute(select(func.coalesce(func.sum(Customer.total_spent), 0)))
    ).scalar_one()

    today = date.today()
    birthdays_this_month = (
        await db.execute(
            select(func.count())
            .select_from(Customer)
            .where(
                Customer.date_of_birth.is_not(None),
                func.extract("month", Customer.date_of_birth) == today.month,
            )
        )
    ).scalar_one()

    return CustomerStatsResponse(
        data=CustomerStats(
            total_customers=total_customers,
            vip_customers=vip_customers,
            birthdays_this_month=birthdays_this_month,
            total_revenue=total_revenue,
        )
    )


@router.get("/upcoming-birthdays", response_model=UpcomingBirthdaysResponse)
async def list_upcoming_birthdays(
    db: AsyncSession = Depends(get_db),
    within_days: int = Query(30, ge=1, le=365),
) -> UpcomingBirthdaysResponse:
    """Customers with a known date of birth falling in the next `within_days`
    days (month/day only, year-independent). Empty until a customer profile
    flow starts collecting date_of_birth — no data is invented here."""
    result = await db.execute(select(Customer).where(Customer.date_of_birth.is_not(None)))
    customers = result.scalars().all()

    today = date.today()
    upcoming = [
        (customer, _days_until_next_birthday(customer.date_of_birth, today))
        for customer in customers
    ]
    upcoming = [pair for pair in upcoming if pair[1] <= within_days]
    upcoming.sort(key=lambda pair: pair[1])

    return UpcomingBirthdaysResponse(
        data=[
            UpcomingBirthday(
                id=customer.public_id,
                name=customer.name,
                is_vip=customer.is_vip,
                date=customer.date_of_birth,
                days_away=days_away,
            )
            for customer, days_away in upcoming
        ]
    )


@router.patch("/{customer_id}", response_model=CustomerCreateResponse)
async def update_customer(
    customer_id: UUID, payload: CustomerUpdate, db: AsyncSession = Depends(get_db)
) -> CustomerCreateResponse:
    customer = await _get_customer_or_404(db, customer_id)
    updates = payload.model_dump(exclude_unset=True)

    if "phone" in updates:
        try:
            normalized_phone = normalize_phone(updates["phone"])
        except InvalidPhoneNumberError as exc:
            raise ValidationAppError(str(exc)) from exc

        conflict = (
            await db.execute(
                select(Customer).where(
                    Customer.normalized_phone == normalized_phone,
                    Customer.id != customer.id,
                )
            )
        ).scalar_one_or_none()
        if conflict is not None:
            raise ValidationAppError("A customer with this phone number already exists")

        customer.normalized_phone = normalized_phone
        customer.phone = updates.pop("phone").strip()

    if "status" in updates:
        valid_statuses = {member.value for member in CustomerStatus}
        if updates["status"] not in valid_statuses:
            raise ValidationAppError(
                f"Invalid status {updates['status']!r}; expected one of {sorted(valid_statuses)}"
            )

    for field, value in updates.items():
        setattr(customer, field, value)

    await db.commit()
    await db.refresh(customer)

    return CustomerCreateResponse(data=CustomerRead.model_validate(customer))


@router.delete("/{customer_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_customer(customer_id: UUID, db: AsyncSession = Depends(get_db)) -> None:
    customer = await _get_customer_or_404(db, customer_id)
    await db.delete(customer)
    await db.commit()


@router.post(
    "/{customer_id}/profile-token",
    response_model=CustomerProfileTokenResponse,
    status_code=status.HTTP_201_CREATED,
)
async def issue_customer_profile_token(
    customer_id: UUID, db: AsyncSession = Depends(get_db)
) -> CustomerProfileTokenResponse:
    """Mints the secure link (`/customer/{token}`) an admin sends a customer
    so they can complete their own profile. Revokes any still-active token
    for this customer first, so at most one link works at a time."""
    customer = await _get_customer_or_404(db, customer_id)

    now = datetime.now(UTC)
    active_tokens = (
        await db.execute(
            select(CustomerProfileToken).where(
                CustomerProfileToken.customer_id == customer.id,
                CustomerProfileToken.revoked_at.is_(None),
                CustomerProfileToken.expires_at > now,
            )
        )
    ).scalars().all()
    for existing_token in active_tokens:
        existing_token.revoked_at = now

    new_token = CustomerProfileToken(customer_id=customer.id)
    db.add(new_token)
    await db.commit()
    await db.refresh(new_token)

    return CustomerProfileTokenResponse(
        data=CustomerProfileTokenIssued(token=new_token.token, expires_at=new_token.expires_at)
    )

File: frontend/lib/api/customers.ts
import {
  API_BASE_URL,
  apiDelete,
  apiGet,
  apiPatch,
  apiPost,
  buildQueryString,
  getAuthorizationHeader,
} from "@/lib/api/client";
import { ApiError, NetworkError } from "@/lib/api/types";
import type { ApiEnvelope, ApiListEnvelope, PaginatedResponse } from "@/lib/api/types";
import type { CustomerTypeOption } from "@/lib/api/customer-types";
import type { Customer, CustomerStatus, CustomerTier } from "@/lib/mock/customers";

export type { Customer, CustomerStatus, CustomerTier } from "@/lib/mock/customers";
export type { CustomerTypeOption } from "@/lib/api/customer-types";
export { formatCurrency } from "@/lib/mock/customers";

export type CustomersSortBy = "name" | "totalSpent" | "totalOrders";
export type SortDirection = "asc" | "desc";

export interface ListCustomersParams {
  page?: number;
  pageSize?: number;
  search?: string;
  status?: CustomerStatus | "all";
  /** Opaque customer type id (from `listCustomerTypes()`), or "all". */
  customerTypeId?: string | "all";
  sortBy?: CustomersSortBy;
  sortDir?: SortDirection;
  /** Only customers verified through at least one campaign. */
  verified?: boolean;
  /** Only customers who've opted into marketing email. */
  marketingOptIn?: boolean;
}

const DEFAULT_PAGE_SIZE = 50;

/** Shape of one item in `GET /api/v1/customers`'s `data` array, after `camelizeKeys`. */
interface CustomerDto {
  id: string;
  name: string;
  phone: string;
  email: string | null;
  address: string | null;
  city: string | null;
  dateOfBirth: string | null;
  isVip: boolean;
  marketingOptIn: boolean;
  customerType: CustomerTypeOption;
  totalSpent: string | number;
  status: string;
  createdAt: string;
}

function getInitials(name: string): string {
  const initials = name
    .split(" ")
    .filter(Boolean)
    .map((part) => part[0])
    .join("")
    .slice(0, 2)
    .toUpperCase();
  return initials || "?";
}

function toStatus(status: string): CustomerStatus {
  switch (status.toLowerCase()) {
    case "inactive":
      return "Inactive";
    case "suspended":
      return "Suspended";
    default:
      return "Active";
  }
}

/** Always rendered in Asia/Dhaka time — the store operates only in
 * Bangladesh, regardless of the server's own system timezone. */
function formatJoinedDate(iso: string): string {
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return "—";
  return date.toLocaleDateString("en-US", {
    month: "short",
    year: "numeric",
    timeZone: "Asia/Dhaka",
  });
}

/** `iso` is a plain "YYYY-MM-DD" date (no time/timezone) — parsed manually
 * so a UTC-vs-local mismatch can't shift the displayed day. */
function formatBirthdayDate(iso: string): string {
  const [, month, day] = iso.split("-").map(Number);
  const date = new Date(Date.UTC(2000, month - 1, day));
  return date.toLocaleDateString("en-US", { month: "short", day: "numeric", timeZone: "UTC" });
}

/**
 * Maps a backend `CustomerDto` onto the UI-facing `Customer` shape every
 * customer component already renders. The backend doesn't track per-order
 * counts or last-purchase timestamps (only aggregate monthly spending), so
 * those fields fall back to neutral placeholders rather than being invented.
 */
function mapDtoToCustomer(dto: CustomerDto): Customer {
  return {
    id: dto.id,
    name: dto.name,
    initials: getInitials(dto.name),
    email: dto.email ?? "No email on file",
    phone: dto.phone,
    address: dto.address,
    city: dto.city,
    tier: dto.isVip ? "VIP" : "Regular",
    status: toStatus(dto.status),
    totalOrders: 0,
    totalSpent: Number(dto.totalSpent),
    joinedAt: formatJoinedDate(dto.createdAt),
    lastPurchaseAt: "—",
    notes: "",
    dateOfBirth: dto.dateOfBirth,
    customerType: dto.customerType,
    marketingOptIn: dto.marketingOptIn,
  };
}

const SORT_BY_TO_BACKEND: Record<CustomersSortBy, string | undefined> = {
  name: "name",
  totalSpent: "total_spent",
  // The backend has no order-count column to sort by; omitting `sort_by`
  // falls back to its default ordering rather than erroring.
  totalOrders: undefined,
};

/** Shared by `listCustomers` and `exportCustomersCsv` — the export must
 * apply the exact same filters as whatever page of results is on screen. */
function buildCustomersFilterQuery(
  params: Omit<ListCustomersParams, "page" | "pageSize">
): Record<string, string | number | boolean | undefined> {
  return {
    search: params.search?.trim() || undefined,
    status: params.status && params.status !== "all" ? params.status.toLowerCase() : undefined,
    customer_type_id:
      params.customerTypeId && params.customerTypeId !== "all"
        ? params.customerTypeId
        : undefined,
    sort_by: params.sortBy ? SORT_BY_TO_BACKEND[params.sortBy] : undefined,
    sort_dir: params.sortDir,
    verified: params.verified,
    marketing_opt_in: params.marketingOptIn,
  };
}

/**
 * Fetches a page of customers from `GET /api/v1/customers`, filtered, sorted,
 * and paginated server-side.
 */
export async function listCustomers(
  params: ListCustomersParams = {}
): Promise<PaginatedResponse<Customer>> {
  const page = params.page ?? 1;
  const pageSize = params.pageSize ?? DEFAULT_PAGE_SIZE;

  const query = buildQueryString({
    page,
    page_size: pageSize,
    ...buildCustomersFilterQuery(params),
  });

  const envelope = await apiGet<ApiListEnvelope<CustomerDto>>(`/customers${query}`);

  return {
    items: envelope.data.map(mapDtoToCustomer),
    total: envelope.meta.total,
    page: envelope.meta.page,
    pageSize: envelope.meta.pageSize,
  };
}

/**
 * Downloads `GET /api/v1/customers/export` (every customer matching the
 * given filters, not just one page) as a CSV file. Bypasses `apiFetch`
 * (JSON-only) for a raw `fetch` the same way `uploadCustomerImport` does for
 * its own non-JSON request — a plain `<a href>` can't carry the
 * `Authorization` header this API requires.
 */
export async function exportCustomersCsv(
  params: Omit<ListCustomersParams, "page" | "pageSize"> = {}
): Promise<void> {
  const query = buildQueryString(buildCustomersFilterQuery(params));

  let response: Response;
  try {
    const authHeader = await getAuthorizationHeader();
    response = await fetch(`${API_BASE_URL}/customers/export${query}`, {
      headers: authHeader,
    });
  } catch (error) {
    throw new NetworkError(error instanceof Error ? error.message : undefined);
  }

  if (!response.ok) {
    const body = await response.text().catch(() => "");
    throw new ApiError(body || response.statusText, response.status);
  }

  const blob = await response.blob();
  const disposition = response.headers.get("Content-Disposition") ?? "";
  const filename = disposition.match(/filename="?([^"]+)"?/)?.[1] ?? "customers-export.csv";

  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(url);
}

export type ProfileStatus = "COMPLETE" | "INCOMPLETE";

/** A row on the POS Customers page (/dashboard/customers/pos) — every
 * customer in the master table, POS-imported or not. Kept separate from
 * `Customer` above (the older, mock-shaped type every other customer
 * component still uses) rather than extending it, so this new page can't
 * accidentally break anything already built on that type. */
export interface PosCustomerRow {
  id: string;
  name: string;
  phone: string;
  customerType: CustomerTypeOption;
  dateOfBirth: string | null;
  address: string | null;
  profileStatus: ProfileStatus;
  totalSpent: number;
  createdAt: string;
}

interface PosCustomerDto {
  id: string;
  name: string;
  phone: string;
  customerType: CustomerTypeOption;
  dateOfBirth: string | null;
  address: string | null;
  profileStatus: ProfileStatus;
  totalSpent: string | number;
  createdAt: string;
}

function mapDtoToPosCustomerRow(dto: PosCustomerDto): PosCustomerRow {
  return {
    id: dto.id,
    name: dto.name,
    phone: dto.phone,
    customerType: dto.customerType,
    dateOfBirth: dto.dateOfBirth,
    address: dto.address,
    profileStatus: dto.profileStatus,
    totalSpent: Number(dto.totalSpent),
    createdAt: dto.createdAt,
  };
}

export interface ListPosCustomersParams {
  page?: number;
  pageSize?: number;
  search?: string;
  customerTypeId?: string | "all";
  profileStatus?: ProfileStatus | "all";
  createdFrom?: string;
  createdTo?: string;
}

/** GET /api/v1/customers with the profile-completeness filter — the same
 * master customer table, just the columns/filters the POS Customers page
 * needs. Server-side filtering and pagination throughout. */
export async function listPosCustomers(
  params: ListPosCustomersParams = {}
): Promise<PaginatedResponse<PosCustomerRow>> {
  const query = buildQueryString({
    page: params.page ?? 1,
    page_size: params.pageSize ?? DEFAULT_PAGE_SIZE,
    search: params.search?.trim() || undefined,
    customer_type_id:
      params.customerTypeId && params.customerTypeId !== "all"
        ? params.customerTypeId
        : undefined,
    profile_status:
      params.profileStatus && params.profileStatus !== "all" ? params.profileStatus : undefined,
    created_from: params.createdFrom,
    created_to: params.createdTo,
  });

  const envelope = await apiGet<ApiListEnvelope<PosCustomerDto>>(`/customers${query}`);

  return {
    items: envelope.data.map(mapDtoToPosCustomerRow),
    total: envelope.meta.total,
    page: envelope.meta.page,
    pageSize: envelope.meta.pageSize,
  };
}

export interface VerifiedCustomerRow {
  id: string;
  name: string;
  phone: string;
  campaignId: string;
  campaignName: string;
  customerType: CustomerTypeOption;
  verifiedAt: string;
  dateOfBirth: string | null;
  address: string | null;
  email: string | null;
}

interface VerifiedCustomerDto {
  id: string;
  name: string;
  phone: string;
  campaignId: string;
  campaignName: string;
  customerType: CustomerTypeOption;
  verifiedAt: string;
  dateOfBirth: string | null;
  address: string | null;
  email: string | null;
}

function mapDtoToVerifiedCustomerRow(dto: VerifiedCustomerDto): VerifiedCustomerRow {
  return {
    id: dto.id,
    name: dto.name,
    phone: dto.phone,
    campaignId: dto.campaignId,
    campaignName: dto.campaignName,
    customerType: dto.customerType,
    verifiedAt: dto.verifiedAt,
    dateOfBirth: dto.dateOfBirth,
    address: dto.address,
    email: dto.email,
  };
}

export interface ListVerifiedCustomersParams {
  page?: number;
  pageSize?: number;
  search?: string;
  campaignId?: string;
  customerTypeId?: string | "all";
  verifiedFrom?: string;
  verifiedTo?: string;
}

/** One row per (customer, campaign) verified pair — a customer verified
 * through two campaigns appears twice. GET /api/v1/customers/verified. */
export async function listVerifiedCustomers(
  params: ListVerifiedCustomersParams = {}
): Promise<PaginatedResponse<VerifiedCustomerRow>> {
  const query = buildQueryString({
    page: params.page ?? 1,
    page_size: params.pageSize ?? DEFAULT_PAGE_SIZE,
    search: params.search?.trim() || undefined,
    campaign_id: params.campaignId,
    customer_type_id:
      params.customerTypeId && params.customerTypeId !== "all"
        ? params.customerTypeId
        : undefined,
    verified_from: params.verifiedFrom,
    verified_to: params.verifiedTo,
  });

  const envelope = await apiGet<ApiListEnvelope<VerifiedCustomerDto>>(`/customers/verified${query}`);

  return {
    items: envelope.data.map(mapDtoToVerifiedCustomerRow),
    total: envelope.meta.total,
    page: envelope.meta.page,
    pageSize: envelope.meta.pageSize,
  };
}

export interface CustomerStats {
  totalCustomers: number;
  vipCustomers: number;
  birthdaysThisMonth: number;
  totalRevenue: number;
}

interface CustomerStatsDto {
  totalCustomers: number;
  vipCustomers: number;
  birthdaysThisMonth: number;
  totalRevenue: string | number;
}

/** Descriptive counts for the dashboard overview, computed from real customer rows. */
export async function getCustomerStats(): Promise<CustomerStats> {
  const envelope = await apiGet<ApiEnvelope<CustomerStatsDto>>("/customers/stats");
  return {
    totalCustomers: envelope.data.totalCustomers,
    vipCustomers: envelope.data.vipCustomers,
    birthdaysThisMonth: envelope.data.birthdaysThisMonth,
    totalRevenue: Number(envelope.data.totalRevenue),
  };
}

export interface RecentCustomer extends Customer {
  /** Raw ISO timestamp, kept alongside the coarse `joinedAt` label so callers
   * that need relative time (e.g. the activity feed) don't have to re-fetch. */
  createdAt: string;
}

/** The most recently added real customers, for the dashboard overview widgets. */
export async function listRecentCustomers(limit: number = 6): Promise<RecentCustomer[]> {
  const query = buildQueryString({
    page: 1,
    page_size: limit,
    sort_by: "created_at",
    sort_dir: "desc",
  });

  const envelope = await apiGet<ApiListEnvelope<CustomerDto>>(`/customers${query}`);
  return envelope.data.map((dto) => ({ ...mapDtoToCustomer(dto), createdAt: dto.createdAt }));
}

export interface UpcomingBirthday {
  id: string;
  name: string;
  initials: string;
  tier: CustomerTier;
  date: string;
  daysAway: number;
}

interface UpcomingBirthdayDto {
  id: string;
  name: string;
  isVip: boolean;
  date: string;
  daysAway: number;
}

/** Customers with a known date of birth in the next `withinDays` days. Empty
 * until a customer profile flow starts collecting date of birth. */
export async function listUpcomingBirthdays(withinDays: number = 30): Promise<UpcomingBirthday[]> {
  const query = buildQueryString({ within_days: withinDays });
  const envelope = await apiGet<ApiListEnvelope<UpcomingBirthdayDto>>(
    `/customers/upcoming-birthdays${query}`
  );

  return envelope.data.map((dto) => ({
    id: dto.id,
    name: dto.name,
    initials: getInitials(dto.name),
    tier: dto.isVip ? "VIP" : "Regular",
    date: formatBirthdayDate(dto.date),
    daysAway: dto.daysAway,
  }));
}

export interface CreateCustomerInput {
  name: string;
  phone: string;
  email?: string;
  address?: string;
  city?: string;
  /** "YYYY-MM-DD", e.g. straight from an `<input type="date">`. */
  dateOfBirth?: string;
  isVip?: boolean;
  marketingOptIn?: boolean;
  /** Omitted defaults to the built-in "General" type server-side. */
  customerTypeId?: string;
}

/** Creates a real customer row via `POST /api/v1/customers`. Throws `ApiError`
 * (e.g. invalid/duplicate phone, 422) or `NetworkError` on failure. */
export async function createCustomer(input: CreateCustomerInput): Promise<Customer> {
  const envelope = await apiPost<ApiEnvelope<CustomerDto>>("/customers", {
    name: input.name,
    phone: input.phone,
    email: input.email,
    address: input.address,
    city: input.city,
    date_of_birth: input.dateOfBirth,
    is_vip: input.isVip ?? false,
    marketing_opt_in: input.marketingOptIn ?? false,
    customer_type_id: input.customerTypeId,
  });

  return mapDtoToCustomer(envelope.data);
}

export interface UpdateCustomerInput {
  name?: string;
  phone?: string;
  /** `null` clears the field; `undefined` leaves it unchanged. */
  email?: string | null;
  address?: string | null;
  city?: string | null;
  dateOfBirth?: string | null;
  isVip?: boolean;
  marketingOptIn?: boolean;
  status?: CustomerStatus;
  customerTypeId?: string;
}

/** Updates a real customer row via `PATCH /api/v1/customers/{id}`. Only the
 * fields present on `input` are sent, matching the backend's partial-update
 * semantics. Throws `ApiError` (e.g. invalid/duplicate phone, 422) or
 * `NetworkError` on failure. */
export async function updateCustomer(id: string, input: UpdateCustomerInput): Promise<Customer> {
  const body: Record<string, unknown> = {};
  if (input.name !== undefined) body.name = input.name;
  if (input.phone !== undefined) body.phone = input.phone;
  if (input.email !== undefined) body.email = input.email;
  if (input.address !== undefined) body.address = input.address;
  if (input.city !== undefined) body.city = input.city;
  if (input.dateOfBirth !== undefined) body.date_of_birth = input.dateOfBirth;
  if (input.isVip !== undefined) body.is_vip = input.isVip;
  if (input.marketingOptIn !== undefined) body.marketing_opt_in = input.marketingOptIn;
  if (input.status !== undefined) body.status = input.status.toLowerCase();
  if (input.customerTypeId !== undefined) body.customer_type_id = input.customerTypeId;

  const envelope = await apiPatch<ApiEnvelope<CustomerDto>>(`/customers/${id}`, body);
  return mapDtoToCustomer(envelope.data);
}

/** Deletes a real customer row (and their spending history, via DB cascade)
 * via `DELETE /api/v1/customers/{id}`. */
export async function deleteCustomer(id: string): Promise<void> {
  await apiDelete<void>(`/customers/${id}`);
}
